incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java

```diff
@@ -15,6 +15,4122 @@
+        int msgTypeId,
+        DirectBuffer buffer,
+        int index,
+        int length,
+        MessageConsumer sender)
+    {
+        final BeginFW begin = beginRO.wrap(buffer, index, index + length);
+        final long originId = begin.originId();
+        final long routedId = begin.routedId();
+        final long initialId = begin.streamId();
+        final long authorization = begin.authorization();
+
+        MqttBindingConfig binding = bindings.get(routedId);
+
+        final MqttRouteConfig resolved = binding != null ? binding.resolve(authorization) : null;
+
+        MessageConsumer newStream = null;
+
+        if (resolved != null)
+        {
+            final long resolvedId = resolved.id;
+
+            final OctetsFW extension = begin.extension();
+            final ExtensionFW beginEx = extension.get(extensionRO::tryWrap);
+            assert beginEx != null;
+            final int typeId = beginEx.typeId();
+            assert typeId == mqttTypeId;
+
+
+            final MqttBeginExFW mqttBeginEx = extension.get(mqttBeginExRO::tryWrap);
+            String16FW clientId;
+            MqttClient client;
+            switch (mqttBeginEx.kind())
+            {
+            case MqttBeginExFW.KIND_SESSION:
+                clientId = mqttBeginEx.session().clientId();
+                client = resolveClient(routedId, resolvedId, supplyInitialId.applyAsLong(resolvedId), clientId);
+                client.sessionStream = new MqttSessionStream(client, sender, originId, routedId, initialId);
+                newStream = client.sessionStream::onSession;
+                break;
+            case MqttBeginExFW.KIND_PUBLISH:
+                final MqttPublishBeginExFW publishBeginEx = mqttBeginEx.publish();
+                clientId = publishBeginEx.clientId();
+                client = resolveClient(routedId, resolvedId, supplyInitialId.applyAsLong(resolvedId), clientId);
+                MqttPublishStream publishStream = new MqttPublishStream(client, sender, originId, routedId, initialId);
+                newStream = publishStream::onPublish;
+                break;
+            case MqttBeginExFW.KIND_SUBSCRIBE:
+                final MqttSubscribeBeginExFW subscribeBeginEx = mqttBeginEx.subscribe();
+                clientId = subscribeBeginEx.clientId();
+                client = resolveClient(routedId, resolvedId, supplyInitialId.applyAsLong(resolvedId), clientId);
+                MqttSubscribeStream subscribeStream = new MqttSubscribeStream(client, sender, originId, routedId, initialId);
+                newStream = subscribeStream::onSubscribe;
+                break;
+            }
+        }
+
+        return newStream;
+    }
+
+    private MqttClient resolveClient(
+        long routedId,
+        long resolvedId,
+        long initialId,
+        String16FW clientId)
+    {
+        final int clientKey = clientKey(clientId.asString());
+        return clients.computeIfAbsent(clientKey,
```

Comment:
Avoid using `.intern()` on strings that can be provided by external untrusted clients.
Suggest using `Object2ObjectHashMap` for now, and later switch to using `affinity` (see core.idl) instead of passing `clientId` in `publish` and `subscribe` streams extensions.

Reply:
Use affinity, and Int2ObjectHashmap